# Laboratorio 4 - Aprendizaje profundo para la clasificación de imágenes

### Roberto Castillo
### Juan Lorthiois

Siga las instrucciones en negritas para completar el laboratorio.

------------

## El reto

**Su tarea es la de construir un clasificador de imágenes usando Keras (Tensorflow) y Redes Neuronales Convolucionales (CNN) para un conjunto de datos conocido como "Fashion MNIST dataset""
. Este conjunto de datos incluye 10 etiquetas de diferentes tipos de ropa con imágenes de 28 by 28 *escalagris*. Hay un conjunto de datos de entrenamiento de 60,000 imágenes y un conjunto de datos de prueba de 10,000 imágenes.**

    Etiqueta	Descripción
    0	    T-shirt/top
    1	    Trouser
    2	    Pullover
    3	    Dress
    4	    Coat
    5	    Sandal
    6	    Shirt
    7	    Sneaker
    8	    Bag
    9	    Ankle boot
    
 

## Los datos

**Tarea 1: Ejecute el siguiente código para descargar los datos usando Keras.**

In [1]:
from tensorflow.keras.datasets import fashion_mnist

(X_entreno, y_entreno), (X_prueba, y_prueba) = fashion_mnist.load_data()

C:\Users\jlort\miniconda3\envs\tensorflow\lib\site-packages\scipy\__init__.py:146: UserWarning: A NumPy version >=1.16.5 and <1.23.0 is required for this version of SciPy (detected version 1.23.1
  warnings.warn(f"A NumPy version >={np_minversion} and <{np_maxversion}"


## Visualización de los Datos

**TAREA 2: Utilice matplotlib para visualizar una imagen del conjunto de datos.  Puede ser cualquier imagen del conjunto de datos.**

In [5]:
import matplotlib.pyplot as plt
import random 
%matplotlib inline

In [ ]:
sample = random.randint(0, 60000)
image = X_entreno[sample]
fig = plt.figure
plt.imshow(image, cmap='viridis')
plt.show()

## Preprocesamiento de los Datos

**TAREA 3: Normalice los datos X entreno y X prueba dividiendo por el valor máximo de los arreglos de las imágenes.**

In [2]:
print(X_entreno.max())
print(X_prueba.max())

255
255


In [3]:
X_entreno= X_entreno / 255
X_prueba= X_prueba /255

In [4]:
print(X_entreno.max())
print(X_prueba.max())

1.0
1.0


**TAREA 4: Cambie el formato de los arreglos X para que incluyan una 4rta dimensión del canal de color. Similar a lo que se hizo en clase para el conjunto de datos MNIST de números.**

In [5]:
X_entreno.shape

(60000, 28, 28)

In [6]:
X_entreno = X_entreno.reshape(60000, 28, 28, 1)
X_prueba = X_prueba.reshape(10000, 28, 28, 1)

**TAREA 5: Convierta los valores de y_entreno y y_prueba para que estén "one-hot encoded" para poder hacer un análisis categórico con Keras.**

In [7]:
from tensorflow.keras.utils import to_categorical

In [8]:
y_entreno

array([9, 0, 0, ..., 3, 0, 5], dtype=uint8)

In [9]:
y_prueba

array([9, 2, 1, ..., 8, 1, 5], dtype=uint8)

In [10]:
y_categorica_entreno = to_categorical(y_entreno)
y_categorica_prueba = to_categorical(y_prueba)

## Configuración del Modelo

**TAREA 6: Utilice Keras para crear un modelo que contenga, al menos, las siguientes capas (pero siéntase en libertad de experimentar):**

* Capa "2D Convolutional", filtros = 32 y tamaño_kernel = (4, 4)
* Capa de "Pooling""
 de tamaño = (2, 2)

* Capa de Aplanado
* Capa Densa (128 unidades, pero siéntase en libertad de "jugar"con este valor), activación RELU

* Una capa Final Densa de 10 unidades con activación softmax

**Luego compile el modelo con estos parámetros: loss = 'categorical_crossentropy', optimizer = 'rmsprop', metrics = ['accuracy']**

In [11]:
#importamos las librerias necesarias (vistas en clase)

from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Activation, Dropout, Flatten, Dense, Conv2D, MaxPooling2D


In [12]:
# Definimos el modelo
modelo = Sequential()

#Agregamos la capa "2dConvolucional" con 32 filtros y un tamaño de kernel (4,4)
modelo.add(Conv2D(filters = 32, kernel_size = (4, 4), input_shape = (28,28,1)))

#Agregamos la capa de Pooling de tamaño (2,2)
modelo.add(MaxPooling2D(pool_size = (2, 2)))

#Capa de aplanado
modelo.add(Flatten())

#capa densa de 128 unidades
modelo.add(Dense(128, activation = 'relu'))

#Capa final densa de 10 unidades
modelo.add(Dense(10, activation = 'softmax'))

#Compilamos el modelo
modelo.compile(loss = 'categorical_crossentropy', optimizer = 'rmsprop', metrics = ['accuracy'])

In [13]:
#Obtenemos un resumen de los parametros selccionados anteriormente

modelo.summary()

Model: "sequential"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 conv2d (Conv2D)             (None, 25, 25, 32)        544       
                                                                 
 max_pooling2d (MaxPooling2D  (None, 12, 12, 32)       0         
 )                                                               
                                                                 
 flatten (Flatten)           (None, 4608)              0         
                                                                 
 dense (Dense)               (None, 128)               589952    
                                                                 
 dense_1 (Dense)             (None, 10)                1290      
                                                                 
Total params: 591,786
Trainable params: 591,786
Non-trainable params: 0
__________________________________________________

### Entrenamiento del Modelo
**TAREA 6: Entrene/Ajuste el modelo con el conjunto X_entreno set. La cantidad de épocas le queda a Ud determinar.**

In [14]:
#Ajustamos/entrenamos el modelo a los datos
modelo.fit(X_entreno, y_categorica_entreno, epochs = 10, validation_data = (X_prueba, y_categorica_prueba))

Epoch 1/10
1875/1875 [==============================] - 22s 12ms/step - loss: 0.3949 - accuracy: 0.8584 - val_loss: 0.3223 - val_accuracy: 0.8846
Epoch 2/10
1875/1875 [==============================] - 22s 12ms/step - loss: 0.2712 - accuracy: 0.9017 - val_loss: 0.3043 - val_accuracy: 0.8909
Epoch 3/10
1875/1875 [==============================] - 23s 12ms/step - loss: 0.2256 - accuracy: 0.9178 - val_loss: 0.3058 - val_accuracy: 0.8900
Epoch 4/10
1875/1875 [==============================] - 21s 11ms/step - loss: 0.1931 - accuracy: 0.9312 - val_loss: 0.3032 - val_accuracy: 0.8975
Epoch 5/10
1875/1875 [==============================] - 22s 12ms/step - loss: 0.1679 - accuracy: 0.9401 - val_loss: 0.3125 - val_accuracy: 0.9027
Epoch 6/10
1875/1875 [==============================] - 23s 12ms/step - loss: 0.1469 - accuracy: 0.9479 - val_loss: 0.3188 - val_accuracy: 0.8968
Epoch 7/10
1875/1875 [==============================] - 23s 12ms/step - loss: 0.1279 - accuracy: 0.9557 - val_loss: 0.3384 -

### Evaluación del Modelo

**TAREA 7: Muestre los valores de [accuracy, precision, recall, f1-score] que logró el modelo con el conjunto de datos X_prueba data set. Tenga en mente que hay múltiples formas de hacer esto.  Sin embargo, le recomendamos que utilice el mismo procedimiento usado mencionado en la parte de intuición, en clase.**

In [15]:
import numpy as np

In [16]:
#Vemos que contamos con las metricas usuales de accuracy y perdida
modelo.metrics_names

['loss', 'accuracy']

In [17]:
#Importamos las librerias necesarias para evaluar el modelo
from sklearn.metrics import classification_report

In [18]:
#Realizamos predicciones sobre el conjunto de prueba (el formato es un poco distinto al visto en clase por la version de tensorflow empleada)
predicciones = modelo.predict(X_prueba) 

#Obtenemos los datos categoricos
clases=np.argmax(predicciones,axis=1)

In [19]:
#generamos el reporte
print(classification_report(y_prueba, clases))

              precision    recall  f1-score   support

           0       0.89      0.79      0.84      1000
           1       1.00      0.96      0.98      1000
           2       0.87      0.83      0.85      1000
           3       0.88      0.93      0.90      1000
           4       0.83      0.87      0.85      1000
           5       0.99      0.95      0.97      1000
           6       0.70      0.78      0.74      1000
           7       0.93      0.98      0.95      1000
           8       0.98      0.97      0.98      1000
           9       0.96      0.96      0.96      1000

    accuracy                           0.90     10000
   macro avg       0.90      0.90      0.90     10000
weighted avg       0.90      0.90      0.90     10000



### Podemos ver que el f1 score es superior a 0.96 en la utlima iteracion lo cual es considerado "excelente"

### Además, el accuracy al comparar las predicciones con los datos de prueba resulta ser del 0.9 lo cual también es excelente!

# Gran trabajo!